src/Compilers/Core/Portable/EmbeddedText.cs

```diff
@@ -0,0 +1,345 @@
+
+            SourceText.ValidateChecksumAlgorithm(checksumAlgorithm);
+
+            return new EmbeddedText(
+                filePath,
+                SourceText.CalculateChecksum(stream, checksumAlgorithm),
+                checksumAlgorithm,
+                CreateBlob(stream));
+        }
+
+        public static EmbeddedText FromBytes(string filePath, byte[] buffer, int length, SourceHashAlgorithm checksumAlgorithm = SourceHashAlgorithm.Sha1)
+        {
+            if (filePath == null)
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            if (filePath.Length == 0)
+            {
+                throw new ArgumentException(CodeAnalysisResources.ArgumentCannotBeEmpty, nameof(filePath));
+            }
+
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (length < 0 || length > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            SourceText.ValidateChecksumAlgorithm(checksumAlgorithm);
+
+            return new EmbeddedText(
+                filePath, 
+                SourceText.CalculateChecksum(buffer, 0, length, checksumAlgorithm),
+                checksumAlgorithm, 
+                CreateBlob(buffer, length));
+        }
+
+        private EmbeddedText(string filePath, ImmutableArray<byte> checksum, SourceHashAlgorithm checksumAlgorithm, ImmutableArray<byte> blob)
+        {
+            Debug.Assert(filePath?.Length > 0);
+            Debug.Assert(Cci.DebugSourceDocument.IsSupportedAlgorithm(checksumAlgorithm));
+            Debug.Assert(!blob.IsDefault && blob.Length >= sizeof(int));
+
+            FilePath = filePath;
+            Checksum = checksum;
+            ChecksumAlgorithm = checksumAlgorithm;
+            Blob = blob;
+        }
+
+        /// <summary>
+        /// Creates the blob to be saved to the PDB.
+        /// </summary>
+        internal static ImmutableArray<byte> CreateBlob(Stream stream)
+        {
+            Debug.Assert(stream != null);
+            Debug.Assert(stream.CanRead);
+            Debug.Assert(stream.CanSeek);;
+            Debug.Assert(stream.Length <= int.MaxValue);
+
+            stream.Seek(0, SeekOrigin.Begin);
+            int length = (int)stream.Length;
+
+            if (length < CompressionThreshold)
+            {
+                var builder = Cci.PooledBlobBuilder.GetInstance();
+                builder.WriteInt32(0);
+
+                int bytesWritten = builder.TryWriteBytes(stream, length);
+                if (length != bytesWritten)
+                {
+                    throw new EndOfStreamException();
+                }
+
+                return builder.ToImmutableArrayAndFree();
+            }
+            else
+            {
+                Debug.Assert(length > 0);
+                var builder = BlobBuildingStream.GetInstance();
```

Comment:
nit: swap these two lines, so that the assert is closer to the line that requires length > 0